skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+        xtwx = np.block(
+            [
+                [np.array([[k0]]),
+                k1[None, :]],
+                [k1[:, None], k2],
+            ],
+        ) + epsilon * np.eye(d + 1)
+        xtwy = np.vstack([s0[None, :], s1])
+
+        beta = np.linalg.solve(xtwx, xtwy)
+        return cast("NDArrayFloat", beta[0]) # intercept term
+
+    def _mean_lls(
+        self,
+        h: float,
+        t_eval: NDArrayFloat,
+        t_obs: NDArrayFloat,
+        y_obs: NDArrayFloat,
+        kernel: KernelFunction,
+    ) -> NDArrayFloat:
+        """
+        Local linear smoother for mean estimation.
+
+        Args:
+            h: Bandwidth for the kernel.
+            t_eval: Query points where smoother is evaluated.
+            t_obs: Observed time points.
+            y_obs: Observed function values.
+            kernel: Kernel function to use for smoothing.
+
+        Returns:
+            Array with smooth estimates for each query point.
+        """
+        epsilon = 1e-8
+
+        t_eval = np.atleast_2d(t_eval)
+        t_obs = np.atleast_2d(t_obs)
+        y_obs = np.atleast_2d(y_obs)
+
+        n_eval, d = t_eval.shape
+        n_obs, q = y_obs.shape
+
+        # (n_eval, n_obs, d): differences for each eval-obs pair
+        diffs = t_eval[:, None, :] - t_obs[None, :, :]
+
+        # Compute kernel weights
+        weights = kernel(diffs / h)
+
+        estimates = np.empty((n_eval, q))
+
+        for i in range(n_eval):
+            estimates[i] = self._compute_local_estimate(
+                xi=diffs[i],
+                yi=y_obs,
+                wi=weights[i],
+                d=d,
+                epsilon=epsilon,
+            )
+
+        return estimates
+
+    def _collect_raw_covariance_lists(
+        self,
+        points: NDArrayFloat,
+        values: NDArrayFloat,
+        start_indices: NDArrayInt,
+        end_indices: NDArrayInt,
+        time_points: NDArrayFloat,
+        mean: NDArrayFloat,
+    ) -> RawCovarianceLists:
+        """
+        Collect raw covariance components as lists from all trajectories.
+
+        Returns:
+            t_1: List of first time points in each pair.
+            t_2: List of second time points in each pair.
+            raw_cov: List of raw covariance matrices (outer products).
+            subj_idx: List of subject indices.
+        """
+        t_1, t_2, raw_cov, subj_idx = [], [], [], []
+
+        for i, start_i in enumerate(start_indices):
+            p_i = points[start_i:end_indices[i]]
+            v_i = values[start_i:end_indices[i]]
+
+            _, indices = cKDTree(time_points).query(p_i)
+            mean_proj = mean[indices]
+            r_i = v_i - mean_proj
+
+            for j, p_ij in enumerate(p_i):
+                for k, p_ik in enumerate(p_i):
+                    t_1.append([float(p_ij)])
+                    t_2.append([float(p_ik)])
+                    raw_cov.append(r_i[j] * r_i[k])
+                    subj_idx.append(i)
```

Comment:
I do not get the point of doing this with a Python loop, when all these can be given as arrays.